### ISSUE TITLE
Add filter to restrict metrics scraped in prometheus receiver

### ISSUE THREAD

**Issue by rghetia** [open-telemetry/opentelemetry-collector#297]
# Problem Description
A prometheus target may provide many metrics but may be only a subset is desirable. Processing unnecessary metrics causes undue burden on resources in otelsvc pipelline. It also increase the backend cost.

# Solution
The solution is provide Filter capability in the Receiver. 

# Tasks
- [ ] Filter Configuration
- [ ] Filter execution

## Should this be limited to Prometheus Receiver?
This may be applicable to other receivers. However, configuration format may differ. For example, prometheus is a pull model and has a notion of targets. In push model source is not known ahead of time. So filter cannot be per source.

## Should the filter be a processor?
It is certainly an option. Advantage of filter processor is that it can be generic and be applied to all metrics irrespective of source. The downside is the receiver may end up doing additional work of transformation for filtered metrics that eventually get dropped. 

Receiver Filter and Processor Filter can both exist simultaneously and can provide different way to filter.

**Comment by annanay25** [comment]
I can pick this up.

Is there consensus on whether to implement as part of receiver/processor?

**Comment by tigrannajaryan** [comment]
My preference would be to make this a processor, unless it is a huge performance hit to receive and drop in the processor. Some experimentation/benchmarking is needed to make a decision on this.

If it is implemented as a processor we need to see if it can be configured consistently with how for example include/exclude filters are configured in span "attributes" processor.

This definitely requires a design doc and a review of the design before implementation starts because there are many unclear issues.

**Comment by anniefu** [comment]
I was unaware of this Issue, but I implemented a filter processor for Knative community's needs (concerned with metrics and Stackdriver in particular). I was planning on submitting a PR for it soon: https://github.com/anniefu/opentelemetry-collector/commit/0df4188b55bdc5d969eb6ecb9852f739a1a28e3c

A couple of key things that are different from span/attribute processor:
- Only one of include/exclude can be specified, but filter processors can be chained
- Filter can be uniquely configured per trace/metric field to filter on (currently just name)
- Optional LRU caching for regexp mode

I did notice the similar include/exclude filters in the span/attributes processors, and tried my best to make it similar (hence the match_type regexp & strict are both included). However, the way the configuration and implementation had embedded understanding of spans made it difficult to re-use for metrics too.

The filter processor also doesn't need to support both include/exclude since the filtering is the processing vs. the span/attributes processor which filter and then do processing. Users can chain include/exclude filter processors in whatever order they want and run as many filter processors as required to achieve their end set of metrics/traces.

I'm happy to discuss this further, though I'm not sure what the best forum is. Maybe the PR itself?
